### ISSUE TITLE
About training 

### ISSUE THREAD

**Issue by CVUsers** [khdlr/HED-UNet#9]
Hi, I'm from China. After looking at your code, the clock points to the game data. On my side, a picture + a mask (all JPG) directory structure is: images / 1 jpg..... gt/1. Jpg, and you don't need cutting training (it seems that you have done this operation). I don't need it. Can you modify the code of load dataset to make it universal training? In addition, what's the speed of your network? Suppose it's an ordinary GPU; thank

**Comment by khdlr** [comment]
Hey mate, a quick google search for "semantic segmentation data loader" should prove insightful on how to implement a data loader for your use case. You can use any custom dataloader with the code.

The speed of the network depends greatly on your usecase (image resolution, etc.), but in our tests we found it to be ~30% slower than a regular UNet.

Cheers,
Konrad

**Comment by CVUsers** [comment]
I am assuming there is a folder which contains sub-folders image and mask. The filenames are same. This is a template you can modify it and use. All the best.
import torch.utils.data as data

class DataLoaderSegmentation(data.Dataset):
    def __init__(self, folder_path):
        super(DataLoaderSegmentation, self).__init__()
        self.img_files = glob.glob(os.path.join(folder_path,'image','*.png')
        self.mask_files = []
        for img_path in img_files:
             self.mask_files.append(os.path.join(folder_path,'mask',os.path.basename(img_path)) 

    def __getitem__(self, index):
            img_path = self.img_files[index]
            mask_path = self.mask_files[index]
            data = use opencv or pil read image using img_path
            label =use opencv or pil read label  using mask_path
            return torch.from_numpy(data).float(), torch.from_numpy(label).float()

    def __len__(self):
        return len(self.img_files)

Can I use this method to replace your data reading part? You use NPY,

**Comment by khdlr** [comment]
Yes, that looks like it should work.

**Comment by CVUsers** [comment]
Hi, I trained with my data. The picture is 256 wide and high for 3 channels, and GT is 256 wide and high for 1 channel. My mask has only the edge (only the edge is white, and other places are black). I trained thousands of pieces several times. The loss is about - 1000, and the ACC is about 60. The effect is very poor, but your training is very good. There should be no problem in the data reading part. Ask for advice

**Comment by CVUsers** [comment]
I train several pieces of data similar to yours (not TIF). GT is also white to fill the target area, and loss is also negative. The effect is very poor

**Comment by khdlr** [comment]
Well the code assumes that your input is a segmentation mask and not an edge mask. If you want to train directly on edge masks, you will need to change the code.

The loss will only be negative if your input mask has invalid values (either negative ones or ones that are greater than 1), so it might be a good idea to check that (i.e. by calling `print(target.min(), target.max())` before line 137 in `train.py`.

My guess would be that in your data loading code you're reading the white values as 255. Building on the code you posted above, it should help to change the last line like this:

```
def __getitem__(self, index):
    img_path = self.img_files[index]
    mask_path = self.mask_files[index]
    data = use opencv or pil read image using img_path
    label =use opencv or pil read label  using mask_path

    return torch.from_numpy(data).float() / 255, torch.from_numpy(label > 0).float()
```

**Comment by CVUsers** [comment]
print(target.min(), target.max()) ::::    tensor(0.) tensor(255.);
use your code: tensor(0.) tensor(1.)
I' m training，thanks
How many milliseconds does it take to infer a 256 size image?

**Comment by wxm123wxm** [comment]
How to solve this problem “ No module named 'attr' ”

**Comment by CVUsers** [comment]
>你运行哪个py报这个错

**Comment by khdlr** [comment]
> print(target.min(), target.max()) :::: tensor(0.) tensor(255.);

Yep, that was the issue. Glad that's fixed now :)

Re inference time: For batched inference (on a NVidia V100), we measured inference times of `3.3 ms` for the HED-UNet model on tiles of size 256x256 as opposed to `2.8 ms` for a plain UNet. But this will depend a lot on data loading, pre-processing and other stuff of course.

**Comment by wxm123wxm** [comment]
您好，感谢您的回复，是在运行train.py的时候报错No module named 'attr'


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "khdlr/HED-UNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年3月30日(星期三) 下午5:26
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [khdlr/HED-UNet] About training  (Issue #9)





  
你运行哪个py报这个错

**Comment by CVUsers** [comment]
可以加我微信交流:zxx15277368495z



---Original---
From: ***@***.***&gt;
Date: Thu, Mar 31, 2022 08:45 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [khdlr/HED-UNet] About training  (Issue #9)




 您好，感谢您的回复，是在运行train.py的时候报错No module named 'attr' 
 
 
 ------------------&amp;nbsp;原始邮件&amp;nbsp;------------------ 
 发件人:                                                                                                                        "khdlr/HED-UNet"                                                                                    ***@***.***&amp;gt;; 
 发送时间:&amp;nbsp;2022年3月30日(星期三) 下午5:26 
 ***@***.***&amp;gt;; 
 ***@***.******@***.***&amp;gt;; 
 主题:&amp;nbsp;Re: [khdlr/HED-UNet] About training  (Issue #9) 
 
 
 
 
 
    
 你运行哪个py报这个错

**Comment by khdlr** [comment]
The term `attr` [isn't used in this package at all](https://github.com/khdlr/HED-UNet/search?q=attr), so the error is likely due to some broken dependencies. Can you try with a new virtualenv / conda environment?

**Comment by CVUsers** [comment]
Hello, my training data is a white mask on the edge (only the edge). The training effect is good. I wrote a detect code myself to infer the graph of 256256,3, edge_ PRED is correct, but SEG_ PRED is all black, which is a little strange (although my work is document edge extraction + correction). Let me show you my code: in addition, do I need to turn my GT diagram (only one circle of white on the edge) into a mask with all white inside the edge? Will this improve the accuracy?

dict = torch.load(r'D:\cver\Cls\Auto_Classification\hed_unet\HED-UNet\logs\2022-03-30_10-48-08\checkpoints/10.pt', map_location="cpu")
    model.load_state_dict(dict)
    idx = 0
    img = cv2.imread(r'E:\IMG_1546.jpg')
    # img = DataLoader(img,1)

    # print(img.size())
    img = cv2.resize(img, (256, 256))
    cv2.imshow('demo_ori', img)
    img = np.transpose(img, (2, 0, 1))  # [3,224,224]
    img = torch.from_numpy(img).float() / 255
    img = img.unsqueeze(0)
    img = img.to(dev)
    time1 = time.time()
    y_hat, y_hat_levels = model(img)
    # print('model inference time: ', time.time()-time1)
    seg_pred = torch.argmax(y_hat[:, 1:], dim=1)
    edge_pred = (y_hat[:, 0] > 0).float()
    # print(seg_pred)
    # print(edge_pred)

    # seg_pred_image = seg_pred.mul(255).byte() # 取值范围
    seg_pred_image = seg_pred.cpu().numpy().transpose((1, 2, 0)) # 改变数据大小
    seg_pred_image = np.array(seg_pred_image, dtype=np.uint8)
    edge_pred_image = edge_pred.cpu().numpy().transpose((1, 2, 0)) # 改变数据大小

    print('Total time: ', time.time()-time1)

    cv2.imshow('demo1', seg_pred_image)
    # edge_pred_image = edge_pred.mul(255).byte() # 取值范围
    cv2.imshow('demo2', edge_pred_image)
    cv2.waitKey(0)

**Comment by CVUsers** [comment]
Hi training on CUDA device, pytorch1 7.1 GPU version, but the video memory only occupies 227mb. It's strange that the code hasn't been moved. Model To () is correct,

**Comment by CVUsers** [comment]
Hello, I have a full screen target. After generating the mask, it is all white. Go to training. The data of other shapes are good. All white data can not be recognized even if it is a training set;

The segmentation rate of each edge changes gradually, but the segmentation rate of epoch changes gradually.

Finally, the edges of our mask, GT and PRED are jagged, but looking at the data, the jagged mask is not serious. I'll send you a picture later

**Comment by CVUsers** [comment]
---Original---
From: "Konrad ***@***.***&gt;
Date: Wed, Mar 30, 2022 18:04 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [khdlr/HED-UNet] About training  (Issue #9)




  
print(target.min(), target.max()) :::: tensor(0.) tensor(255.);
  
Yep, that was the issue. Glad that's fixed now :)
 
Re inference time: For batched inference (on a NVidia V100), we measured inference times of 3.3 ms for the HED-UNet model on tiles of size 256x256 as opposed to 2.8 ms for a plain UNet. But this will depend a lot on data loading, pre-processing and other stuff of course.

**Comment by khdlr** [comment]
> Hello, I have a full screen target. After generating the mask, it is all white. Go to training. The data of other shapes are good. All white data can not be recognized even if it is a training set;

An white mask will have no edges, so the edge-head becomes useless in that setup I would say.

> The segmentation rate of each edge changes gradually, but the segmentation rate of epoch changes gradually.

Not sure what you mean by that.

> Finally, the edges of our mask, GT and PRED are jagged, but looking at the data, the jagged mask is not serious. I'll send you a picture later

I didn't receive any picture, feel free to contact me at &nbsp;![image](https://user-images.githubusercontent.com/9249838/162147897-c8c2fb99-691e-4170-9f95-4a05a6a44042.png)

**Comment by wxm123wxm** [comment]
Thank you for your reply! The current problem is as followed:





------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "khdlr/HED-UNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年4月7日(星期四) 下午3:49
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [khdlr/HED-UNet] About training  (Issue #9)





  
Hello, I have a full screen target. After generating the mask, it is all white. Go to training. The data of other shapes are good. All white data can not be recognized even if it is a training set;
  
An white mask will have no edges, so the edge-head becomes useless in that setup I would say.
  
The segmentation rate of each edge changes gradually, but the segmentation rate of epoch changes gradually.
  
Not sure what you mean by that.
  
Finally, the edges of our mask, GT and PRED are jagged, but looking at the data, the jagged mask is not serious. I'll send you a picture later
  
I didn't receive any picture, feel free to contact me at &nbsp;

**Comment by CVUsers** [comment]
Did you get the mail？

**Comment by CVUsers** [comment]
How to make the edge branch be learned when it is around the edge of the image
